src/toxcore/dht_new/packet.rs

```diff
@@ -0,0 +1,938 @@
+    SocketAddrV4,
+    SocketAddrV6
+};
+
+use toxcore::dht_new::binary_io::*;
+use toxcore::crypto_core::*;
+
+/// Length in bytes of [`PingReq`](./struct.PingReq.html) and
+/// [`PingResp`](./struct.PingResp.html) when serialized into bytes.
+pub const PING_SIZE: usize = 9;
+
+/** `NatPing` type byte for [`NatPingReq`] and [`NatPingResp`].
+[./struct.PingReq.html] [./struct.PingResp.html]
+*/
+pub const NAT_PING_TYPE: u8 = 0xfe;
+
+/** Length in bytes of NatPings when serialized into bytes.
+*/
+pub const NAT_PING_SIZE: usize = PING_SIZE + 1;
+
+/** Standard DHT packet that encapsulates in the payload
+[`DhtPacketT`](./trait.DhtPacketT.html).
+
+https://zetok.github.io/tox-spec/#dht-packet
+*/
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DhtPacket {
+    /// [`PingReq`](./struct.PingReq.html) structure.
+    PingReq(PingReq),
+    /// [`PingResp`](./struct.PingResp.html) structure.
+    PingResp(PingResp),
+    /// [`GetNodes`](./struct.GetNodes.html) structure.
+    GetNodes(GetNodes),
+    /// [`SendNodes`](./struct.SendNodes.html) structure.
+    SendNodes(SendNodes),
+}
+
+impl ToBytes for DhtPacket {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        match *self {
+            DhtPacket::PingReq(ref p) => p.to_bytes(buf),
+            DhtPacket::PingResp(ref p) => p.to_bytes(buf),
+            DhtPacket::GetNodes(ref p) => p.to_bytes(buf),
+            DhtPacket::SendNodes(ref p) => p.to_bytes(buf),
+        }
+    }
+}
+
+impl FromBytes for DhtPacket {
+    named!(from_bytes<DhtPacket>, alt!(
+        map!(PingReq::from_bytes, DhtPacket::PingReq) |
+        map!(PingResp::from_bytes, DhtPacket::PingResp) |
+        map!(GetNodes::from_bytes, DhtPacket::GetNodes) |
+        map!(SendNodes::from_bytes, DhtPacket::SendNodes)
+    ));
+}
+
+/** DHT Request packet.
+
+https://zetok.github.io/tox-spec/#dht-request-packets
+*/
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DhtRequest {
+    /// [`NatPingReq`](./struct.NatPingReq.html) structure.
+    NatPingReq(PingReq),
+    /// [`NatPingResp`](./struct.NatPingResp.html) structure.
+    NatPingResp(PingResp),
+}
+
+/** NatPing request of DHT Request packet.
+*/
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct NatPingReq(pub PingReq);
+
+/** NatPing response of DHT Request packet.
+*/
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct NatPingResp(pub PingResp);
+
+impl ToBytes for DhtRequest {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        match *self {
+            DhtRequest::NatPingReq(ref p) => p.to_bytes(buf),
+            DhtRequest::NatPingResp(ref p) => p.to_bytes(buf),
+        }
+    }
+}
+
+impl FromBytes for DhtRequest {
+    named!(from_bytes<DhtRequest>, alt!(
+        map!(PingReq::from_bytes, DhtRequest::NatPingReq) |
+        map!(PingResp::from_bytes, DhtRequest::NatPingResp)
+    ));
+}
+
+impl FromBytes for NatPingReq {
```

Comment:
You have to define struct and then `FromBytes` and `ToBytes` impl. Don't mixup the code, structurize it

Reply:
Mixing code intended to avoid redundancy.
I'm not convinced that what method is better.
Anyway, I accept your opinion.